SimPEG/directives/pgi_directives.py

```diff
@@ -27,108 +26,142 @@
+        ].T
+
+        # Update the GMM parameters if required
+        gmmref = self.pgi_regularization.gmmref
+        if self.update_gmm:
+            # Determine the class of the new gmm
+            if isinstance(gmmref, GaussianMixtureWithNonlinearRelationships):
+                new_gmm_class = GaussianMixtureWithNonlinearRelationshipsWithPrior
+            elif isinstance(gmmref, WeightedGaussianMixture):
+                new_gmm_class = GaussianMixtureWithPrior
             else:
-                clfupdate = copy.deepcopy(self.pgi_reg.gmmref)
-
-            self.pgi_reg.gmm = clfupdate
-            membership = self.pgi_reg.gmm.predict(model)
-
-            if self.fixed_membership is not None:
-                membership[self.fixed_membership[:, 0]] = self.fixed_membership[:, 1]
+                raise ValueError(
+                    f"Invalid class '{gmmref.__class__}' for the Gaussian Mixture"
+                    " in the PGI regularization."
+                )
+            new_gmm = new_gmm_class(
+                gmmref=gmmref,
+                zeta=self.zeta,
+                kappa=self.kappa,
+                nu=self.nu,
+                verbose=self.verbose,
+                prior_type="semi",
+                update_covariances=self.update_covariances,
+                max_iter=self.pgi_regularization.gmm.max_iter,
+                n_init=self.pgi_regularization.gmm.n_init,
+                reg_covar=self.pgi_regularization.gmm.reg_covar,
+                weights_init=self.pgi_regularization.gmm.weights_,
+                means_init=self.pgi_regularization.gmm.means_,
+                precisions_init=self.pgi_regularization.gmm.precisions_,
+                random_state=self.pgi_regularization.gmm.random_state,
+                tol=self.pgi_regularization.gmm.tol,
+                verbose_interval=self.pgi_regularization.gmm.verbose_interval,
+                warm_start=self.pgi_regularization.gmm.warm_start,
+                fixed_membership=self.fixed_membership,
+            )
+            new_gmm = new_gmm.fit(model)
+        else:
+            new_gmm = copy.deepcopy(gmmref)
+
+        # Update the GMM in the PGI regularization
+        self.pgi_regularization.gmm = new_gmm
+        # Compute membership of the current model
+        membership = self.pgi_regularization.gmm.predict(model)
+        # Override the membership if some cells have fixed membership
+        if self.fixed_membership is not None:
+            membership[self.fixed_membership[:, 0]] = self.fixed_membership[:, 1]
+        # Get the new reference model
+        mref = mkvc(self.pgi_regularization.gmm.means_[membership])
+        # Update the reference model in the PGI regularization
+        self.pgi_regularization.reference_model = mref
+        if getattr(self.fixed_membership, "shape", [0, 0])[0] < len(membership):
+            self.pgi_regularization._r_second_deriv = None
```

Comment:
I'm not entirely sure what's the purpose of these lines... I'll keep looking.

Reply:
@santisoler : it is to save the derivatives of the polynomial mapping such as shown in this example: https://docs.simpeg.xyz/content/examples/10-pgi/plot_inv_1_PGI_Linear_1D_joint_WithRelationships.html#sphx-glr-content-examples-10-pgi-plot-inv-1-pgi-linear-1d-joint-withrelationships-py

The Polynom class takes for some unknown reason a very long time to compute derivatives. Thus, instead of computing them on-call, these hidden properties saves them.